_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,205 @@
+  [minimizes the regret bound][minregret] over Mahalanobis norms for projected
+  gradient descent. That said, in real life we just consider the diagonal matrix
+  $diag(A_t)$ for the sake of computation.
+* $\eta_0$ is the initial learning step.
+* $\lambda_1$ is the strength of the lasso regularization.
+* $\lambda_2$ is the strength of the ridge regularization.
+
+Regularization terms aside, this is mostly equivalent to an online gradient
+descent formulation with learning rate $\eta_t = \eta_0 A_t^{-1}$, but solving
+the minimization in a follow-the-leader fashion allows for an effective
+implementation of lasso regularization that produces coefficients which are
+exactly zero (and, hence, more sparse models that are cheaper to store and
+transfer and, most importantly, way faster to evaluate). Although at first sight
+it might seem that implementing the follow-the-leader update step ---which sums
+over every previous input--- would be harder than implementing the gradient
+descent update step ---which just cares about the last update--- this first
+impression turns out to be wrong after a [careful reformulation][trenches] of
+the above expression:
+$$
+\theta_{t+1} = \underset{\theta}{\operatorname{argmin}}\ 
+{z_t}^T \cdot \theta +
+\frac{1}{2\eta_0} ||\theta||^2_{A_t} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where $z_t = {g_{1:t}} - \frac{1}{\eta_0} \sum_{s=1}^t (A_s - A_{s-1})\theta_s$,
+which can be cheaply calculated in an incremental way as:
+$$
+z_t = z_{t-1} + g_t + \frac{1}{\eta_0} (A_t - A_{t-1})\theta_t
+$$
+Now, when $A_t$ is diagonal the closed form of the solution is just:
+$$
+\theta_{t,i} =
+\begin{cases}
+0 & \text{if } |z_{t,i}| \le \lambda_1 \\
+-\frac{z_{t,i} - \text{sgn}(z_{t,i})\lambda_1}{A_{t,i}/\eta_0 - \lambda_2} & \text{otherwise}
+\end{cases}
+$$
+
+Glossing over many technicalities like:
+* Keeping a huge inverse hash map, in order to recover features from hashes.
+* Following a "memory schedule" that weights each input according to the time
+  elapsed since its arrival, in order to adapt to ever-changing market
+  conditions[^decay].
+* Storing frequent checkpoints of the model, in order to resume from valid
+  states after expected deployments or unexpected crashes.
+
+...our implementation faithfully follows the update formula above. Given that
+the input vectors are sparse, the update step is very efficient (it's $O(n)$
+with $n$ the number of non zero features) even for high dimensional data
+(millions of features). The implementation is pretty generic in the sense that
+---besides the logloss and logit link--- custom loss and link functions can be
+passed as parameters to the optimizer. We implemented all this using Python
+(with patches of C here and there) and we are very happy with our experience and
+results.
+
+Currently, we are well on our way to leverage this technology for the estimation
+of models of market clearing prices and campaign velocity of spend, which are
```

Comment:
how about... campaign velocity of spend --> "campaign spend velocity" or "campaign spend speed" ?